### PR TITLE
Implement min/max for GCONSALE

### PR DESCRIPTION
**Pull request by totto82**
Only default action RATE is implemented for violation of maximum sales target

Tested on 9_3D and 9_3E.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by alfbr** [comment]
I can confirm that this PR resolves the 9_3D and 9_3E cases in opm-tests. I will prepare a PR for putting them under continuous testing.

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 1]
```diff
@@ -2246,7 +2246,13 @@ namespace Opm
         std::vector<double> resv_coeff(phaseUsage().num_phases, 1.0);
         rateConverter_.calcCoeff(0, pvtRegionIdx_, resv_coeff); // FIPNUM region 0 here, should use FIPNUM from WELSPECS.
 
-        WellGroupHelpers::TargetCalculator tcalc(currentGroupControl, pu, resv_coeff);
+        // gconsale may adjust the grat target.
+        // the adjusted rates is send to the targetCalculator
+        double gratTargetFromSales = 0.0;
+        if(well_state.hasGroupGratTargetFromSales(group.name()))
```
Comment (atgeirr) [inline opm/simulators/wells/WellInterface_impl.hpp:2252]: Nitpick: space after if, braces. (Here and elsewhere).

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 2]
```diff
@@ -2063,38 +2064,104 @@ namespace Opm {
...
+                OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "GCONSALE exceed limit MAXR not implemented", deferred_logger);
+                break;
+            }
+            case GConSale::MaxProcedure::END: {
+                OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "GCONSALE exceed limit END not implemented", deferred_logger);
+                break;
+            }
+            case GConSale::MaxProcedure::RATE: {
+                    well_state.setCurrentProductionGroupControl(group.name(), Group::ProductionCMode::GRAT);
+                    ss << "Maximum GCONSALE limit violated for " << group.name() << ". The group is switched from " << Group::ProductionCMode2String(oldProductionControl) << " to ";
```
Comment (atgeirr) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:2133]: Nitpick: break long lines. (Very hard to read on GitHub...)

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 3]
```diff
@@ -2063,38 +2064,104 @@ namespace Opm {
...
+            const Group::ProductionCMode& currentProductionControl = well_state.currentProductionGroupControl(group.name());
+            if ( currentProductionControl == Group::ProductionCMode::GRAT ) {
+                ss << "Group " + group.name() + " has sale rate less then minimum permitted value and is under GRAT control. \n";
+                ss << "The GRAT is increased to meet the sales minimum rate. \n";
+                well_state.setCurrentGroupGratTargetFromSales(group.name(), production_target);
+            //} else if () {//TODO add action for WGASPROD
+            //} else if () {//TODO add action for drilling queue
+            } else {
+                ss << "Group " + group.name() + " has sale rate less then minimum permitted value but cannot increase the group production rate, \n";
+                ss << "adjust gas production using WGASPROD nor drill new wells to meat the sales target. No action is taken. \n ";
```
Comment (atgeirr) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:2152]: Do we support WGASPROD? Also, 'nor'->'or' and 'meat'->'meet'.

Comment (totto82) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:2152]: We don't support WGASPROD.

Comment (atgeirr) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:2152]: Then I think we should not recommend it as a solution?

Comment (totto82) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:2152]: It is , not a . So the meaning of the text is that we can not adjust gas production using WGASPROD, etc. I can add something like. "Not supported in Flow. " to clearly this.

Comment (atgeirr) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:2152]: Ah, I misunderstood a little then. Good to have a clearer message.

**Comment by totto82** [comment]
jenkins build this please

**Pull request merged**